### PR TITLE
chore: enhance parallel test execution

### PR DESCRIPTION
**Pull request by adeldhis2**
### Key Changes:
**Conditional Recording Triggers**: Implemented conditional recording triggers via commit messages and GitHub labels to ensure that Cypress tests are only recorded when necessary.
**Parallel Execution Strategies**: Refined parallel execution strategies to improve efficiency and resolve test redundancies in GitHub Actions.
**Debugging Enhancements**: Added detailed debugging outputs to help track the dynamic values of the test environment variables and ensure correct parameter passing in the workflow.
**Update Documentation**: Updated the documentation to reflect the changes and ensure clarity on the new CI/CD processes.

### Expected Outcomes:
**Reduced Cypress Cloud Credit Usage**: By optimising the conditions under which tests are recorded and uploaded to Cypress Cloud, we anticipate a significant reduction in credit usage.
**Increased CI Efficiency**: With better management of parallel execution and conditional testing, the CI process should be more efficient, leading to quicker test cycles.

### Relevant Links:
**GitHub Actions Run**: [View Run](https://github.com/dhis2/line-listing-app/actions/runs/9280013141?pr=514)
**Cypress Cloud Run**: [View Run](https://github.com/dhis2/line-listing-app/actions/runs/9280321181?pr=514)

### Jira Ticket:
For more details, please refer to the Jira ticket: [DHIS2-17373](https://dhis2.atlassian.net/browse/DHIS2-17373).

[DHIS2-17373]: https://dhis2.atlassian.net/browse/DHIS2-17373?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by jenniferarnesen** [comment]
@adeldhis2 Could you make the following comment more specific?

"Parallel Execution Strategies: Refined parallel execution strategies to improve efficiency and resolve test redundancies in GitHub Actions.". 

It would be nice to have some information about what exactly the implementation. Is doing. Looks like you are manually setting up the test groups rather than leaving it to Cypress to divide up the tests. An explanation as to why this would be helpful for future reference.

File: cypress/support/generateTestMatrix.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const fs = require('fs')
```
Comment (jenniferarnesen) [inline cypress/support/generateTestMatrix.js:1]: Please add a jest test for this function.

**Comment by jenniferarnesen** [comment]
In the maps-app PR https://github.com/dhis2/maps-app/pull/3201 you've added info to the README. Can you do the same on this PR?

**Comment by adeldhis2** [comment]
@jenniferarnesen Thank you for the query! Here's why we manually define test groups in our GitHub Actions setup:

**Manual Grouping of Tests**: In GitHub Actions, without specified test groups, all tests would run in every parallel container, leading to unnecessary repetition. By organizing tests into predefined groups, each group runs only once. This prevents redundancy and expedites the testing process. More about [GitHub Actions matrix strategy](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs).

**Why Not Rely on Cypress's Automatic Balancing?** While Cypress automatically assigns tests to machines to balance workload (more on [Cypress load balancing](https://docs.cypress.io/guides/cloud/smart-orchestration/load-balancing)), GitHub Actions requires explicit distribution of tasks. This manual grouping ensures efficient resource usage and prevents multiple executions of the same tests across different runners.

**Benefits**: This approach allows precise control over test distribution, enhancing predictability and efficiency. It reduces total execution time and operational costs by ensuring each test is run precisely once across the available infrastructure.

This method optimises our use of GitHub Actions, ensuring a faster, cost-effective, and efficient testing process.

**Pull request merged**